File: plugins/plugin-code/python/elizaos_plugin_eliza_coder/path_utils.py
from __future__ import annotations

import os
from pathlib import Path

DEFAULT_FORBIDDEN_COMMANDS: list[str] = [
    "rm -rf /",
    "rm -rf ~",
    "sudo rm",
    "mkfs",
    "dd if=/dev",
]


def extract_base_command(command: str) -> str:
    trimmed = command.strip()
    if not trimmed:
        return ""
    return trimmed.split()[0]


def is_safe_command(command: str) -> bool:
    c = command.strip()
    if not c:
        return False
    if "&&" in c or "||" in c or ";" in c:
        return False
    if "$(" in c or "`" in c:
        return False
    return True


def is_forbidden_command(command: str, additional_forbidden: list[str]) -> bool:
    lower = command.lower()
    for f in DEFAULT_FORBIDDEN_COMMANDS:
        if f.lower() in lower:
            return True
    for f in additional_forbidden:
        if f.strip() and f.lower() in lower:
            return True
    return False


def validate_path(target_path: str, allowed_directory: str, current_directory: str) -> str | None:
    allowed = Path(allowed_directory).resolve()
    base = Path(current_directory).resolve() if current_directory else allowed
    resolved = (base / target_path).resolve()

    try:
        resolved.relative_to(allowed)
    except ValueError:
        return None

    return os.fspath(resolved)

File: plugins/plugin-code/typescript/utils/pathUtils.ts
import * as path from "node:path";

export const DEFAULT_FORBIDDEN_COMMANDS: string[] = [
  "rm -rf /",
  "rm -rf ~",
  "sudo rm",
  "mkfs",
  "dd if=/dev",
];

export function extractBaseCommand(command: string): string {
  const trimmed = command.trim();
  if (!trimmed) return "";
  const first = trimmed.split(/\s+/)[0];
  return first ?? "";
}

/**
 * Very conservative “safe command” filter. Blocks shell control operators and
 * common escape patterns. (This is aligned with other plugins’ patterns.)
 */
export function isSafeCommand(command: string): boolean {
  const c = command.trim();
  if (!c) return false;
  if (c.includes("&&") || c.includes("||") || c.includes(";")) return false;
  if (c.includes("$(") || c.includes("`")) return false;
  return true;
}

export function isForbiddenCommand(
  command: string,
  additionalForbidden: string[],
): boolean {
  const c = command.toLowerCase();
  for (const f of DEFAULT_FORBIDDEN_COMMANDS) {
    if (c.includes(f.toLowerCase())) return true;
  }
  for (const f of additionalForbidden) {
    if (f.trim().length === 0) continue;
    if (c.includes(f.toLowerCase())) return true;
  }
  return false;
}

/**
 * Resolve a target path against the current directory while enforcing that
 * the result stays inside allowedDirectory.
 */
export function validatePath(
  targetPath: string,
  allowedDirectory: string,
  currentDirectory: string,
): string | null {
  const base =
    currentDirectory && currentDirectory.length > 0
      ? currentDirectory
      : allowedDirectory;
  const resolved = path.resolve(base, targetPath);
  const allowed = path.resolve(allowedDirectory);

  const rel = path.relative(allowed, resolved);
  if (rel === "") return resolved;
  if (rel.startsWith("..") || path.isAbsolute(rel)) return null;
  return resolved;
}
